app/models/bitbucket_integration.rb

```diff
@@ -0,0 +1,379 @@
+    source_id: :id,
+    number: :id,
+    title: :title,
+    body: :description,
+    url: [:links, :html, :href],
+    state: :state,
+    head_ref: [:source, :branch, :name],
+    base_ref: [:destination, :branch, :name],
+    opened_at: :created_on
+  }
+
+  COMMITS_TRANSFORMATIONS = {
+    url: [:links, :html, :href],
+    commit_hash: :hash,
+    message: :message,
+    author_raw: [:author, :raw],
+    author_login: [:author, :user, :nickname],
+    author_url: [:author, :links, :html, :href],
+    timestamp: :date,
+    parents: {
+      parents: {
+        url: [:links, :html, :href],
+        sha: [:hash]
+      }
+    }
+  }
+
+  def pr_closed?(pr)
+    %w[MERGED DECLINED].include?(pr[:state])
+  end
+
+  def pr_open?(pr)
+    pr[:state] == "OPEN"
+  end
+
+  def repos(workspace)
+    with_api_retries { installation.list_repos(workspace, REPOS_TRANSFORMATIONS) }
+  end
+
+  def find_or_create_webhook!(id:, train_id:)
+    GitHub::Result.new do
+      if id
+        webhook = with_api_retries { installation.find_webhook(code_repository_name, id, WEBHOOK_TRANSFORMATIONS) }
+        if webhook[:url] == events_url(train_id:) && (installation.class::WEBHOOK_EVENTS - webhook[:events]).empty?
+          webhook
+        else
+          with_api_retries { installation.update_repo_webhook!(code_repository_name, webhook[:id], events_url(train_id:), WEBHOOK_TRANSFORMATIONS) }
+        end
+      else
+        with_api_retries { installation.create_repo_webhook!(code_repository_name, events_url(train_id:), WEBHOOK_TRANSFORMATIONS) }
+      end
+    rescue Installations::Bitbucket::Error => ex
+      raise ex unless ex.reason == :webhook_not_found
+      with_api_retries { installation.create_repo_webhook!(code_repository_name, events_url(train_id:), WEBHOOK_TRANSFORMATIONS) }
+    end
+  end
+
+  def create_branch!(from, to, source_type: :branch)
+    with_api_retries { installation.create_branch!(code_repository_name, from, to, source_type:) }
+  end
+
+  def commit_log(from_branch, to_branch)
+    installation.commits_between(code_repository_name, from_branch, to_branch, COMMITS_TRANSFORMATIONS)
+  end
+
+  def diff_between?(from_branch, to_branch, from_type: :branch)
+    with_api_retries { installation.diff?(code_repository_name, from_branch, to_branch, from_type) }
+  end
+
+  def branch_head_sha(branch, sha_only: true)
+    with_api_retries { installation.head(code_repository_name, branch, sha_only:, commit_transforms: COMMITS_TRANSFORMATIONS) }
+  end
+
+  def branch_exists?(branch_name)
+    with_api_retries { installation.get_branch(code_repository_name, branch_name) }.present?
+  rescue Installations::Error => ex
+    raise ex unless ex.reason == :not_found
+    false
+  end
+
+  def branch_url(branch_name)
+    "https://bitbucket.org/#{code_repository_name}/branch/#{branch_name}"
+  end
+
+  def tag_url(tag_name)
+    "https://bitbucket.org/#{code_repository_name}/src/#{tag_name}"
+  end
+
+  def compare_url(to_branch, from_branch)
```

Comment:
Something like this

https://bitbucket.org/tramline/ueno/branch/r/bb-fuill-train/2024-09-16?dest=r%2Fbb-fuill-train%2F2024-09-16